src/ctia/task/migration/migrations.clj

```diff
@@ -149,6 +150,85 @@
                (update :incident_time simplify-incident-time))
            doc))))
 
+;; Add new investigation fields
+
+(defn transient-id? [id]
+  (and (string? id)
+       (re-find #"\Atransient" (str id))))
+
+(defn object-id [doc]
+  (let [id (get doc :id)]
+    (when (not (transient-id? id))
+      id)))
+
+(defn read-actions [investigation]
+  (let [actions (get investigation :actions)]
+    (if (string? actions)
+      (json/parse-string actions true)
+      [])))
+
+(def empty-action-data
+  {:object_ids #{}
+   :investigated_observables #{}
+   :targets #{}})
+
+(defn merge-action-data
+  ([] empty-action-data)
+  ([m] m)
+  ([m1 m2]
+   (merge m1 {:object_ids (into (get m1 :object_ids)
+                                (get m2 :object_ids))
+              :investigated_observables (into (get m1 :investigated_observables)
+                                              (get m2 :investigated_observables))
+              :targets (into (get m1 :targets)
+                             (get m2 :targets))})))
+
+(defmulti derive-action-data
+  (fn [action]
+    (get action :type))
+  :default ::default)
+
+(defmethod derive-action-data ::default [_]
```

Comment:
"Never use `defmethod`"
https://github.com/threatgrid/ctia/blob/master/CONTRIBUTING.org#guidelines
The underlying idea is that it's hard to follow and debug, it can be gracefully replaced by a `case` here that redirect to proper function.

Reply:
Absolutely! I also prefer `case` to `defmethod` exactly for that reason. 😄